src/varname.jl

```diff
@@ -334,37 +342,125 @@ macro vinds(expr::Union{Expr, Symbol})
+    _replace_ref_begin_end(ex::Symbol, withex) = Base.replace_ref_end_!(ex, withex)
+    _index_replacement_for(s) = :($lastindex($s))
+    _index_replacement_for(s, n) = :($lastindex($s, $n))
+else
+    _replace_ref_begin_end(ex, withex) = Base.replace_ref_begin_end_!(copy(ex), withex)
+    _replace_ref_begin_end(ex::Symbol, withex) = Base.replace_ref_begin_end_!(ex, withex)
+    _index_replacement_for(s) = :($firstindex($s)), :($lastindex($s))
+    _index_replacement_for(s, n) = :($firstindex($s, $n)), :($lastindex($s, $n))
+end
+
+
 """
     vinds(expr)
 
 Return the indexing part of the [`@varname`](@ref)-compatible expression `expr` as an expression
-suitable for input of the [`VarName`](@ref) constructor.
+suitable for input of the [`VarName`](@ref) constructor (i.e., a tuple of tuples).
 
 ## Examples
 
 ```jldoctest
-julia> vinds(:(x[end]))
-:((((lastindex)(x),),))
+julia> x = [10, 20, 30]; eval(vinds(:(x[end])))
+((3,),)
+
+
+julia> x = [10 20]; eval(vinds(:(x[1, end])))
+((1, 2),)
+
+julia> x = [[1, 2]]; eval(vinds(:(x[1][end])))
+((1,), (2,))
+
+julia> x = [fill([[10], [20, 30]], 2, 2, 2)]
+       if VERSION < v"1.5.0-DEV.666"
+            eval(vinds(:(x[1][2, end, :][2][end])))
+       else
+            eval(vinds(Meta.parse("x[begin][2, end, :][2][end]")))
+       end
+((1,), (2, 2, Colon()), (2,), (2,))
 
-julia> vinds(:(x[1, end]))
-:(((1, (lastindex)(x, 2)),))
 ```
 """
-function vinds end
+function vinds(expr, head = vsym(expr))
+    # see https://github.com/JuliaLang/julia/blob/bb5b98e72a151c41471d8cc14cacb495d647fb7f/base/views.jl#L17-L75
+    indexing = _straighten_indexing(expr)
+    inds = Expr[]  # collection of result indices
+    partial = head  # partial :ref expressions, used in caching
+    cached_exprs = Vector{Pair{Symbol, Expr}}()  # cache for partial expressions in a :let
+    
+    for ixs in indexing
+        # S becomes the name of the cached variable
+        S = (partial == head) ? head : gensym(:S)
+        used_S = false
+        
+        nixs = length(ixs)
+        if nixs == 1
+            # for 1D indexing, just use `lastindex(x)`
+            ixs[1], used = _replace_ref_begin_end(ixs[1], _index_replacement_for(S))
+            used_S |= used
+        elseif nixs > 1
+            # otherwise, we need `lastindex(x, i)`
+            for i in eachindex(ixs)
+                ixs[i], used = _replace_ref_begin_end(ixs[i], _index_replacement_for(S, i))
+                used_S |= used
+            end
+        end
 
-vinds(expr::Symbol) = Expr(:tuple)
-function vinds(expr::Expr)
-    if Meta.isexpr(expr, :ref)
-        ex = copy(expr)
-        @static if VERSION < v"1.5.0-DEV.666"
-            Base.replace_ref_end!(ex)
+        if used_S && partial !== head
+            # cache that expression if we actually used it, and use the new name in the
+            # partial expression
+            push!(cached_exprs, S => partial)
+            partial = Expr(:ref, S, ixs...)
```

Comment:
Should this be a `view` instead? This would avoid unnecessary allocations (at least on more recent versions of Julia).

EDIT: I guess one issue here is that it will exclude indexing of non-array-types, e.g. `Tuple`:
```julia
julia> x = ((rand(1, 2, 3), ), );

julia> eval(vinds(:(x[1][1][1,end,1])))
ERROR: MethodError: no method matching view(::Tuple{Tuple{Array{Float64, 3}}}, ::Int64)
Closest candidates are:
  view(::AbstractArray, ::Any...) where N at subarray.jl:174
```

So instead we should use `Base.maybeview`.

Reply:
I had tried a version with `view` here, but the main problem was that it returns 0-D views for scalar indices.  But I didn't know about `maybeview`, that's awesome!